### PR TITLE
Dynamic add tSafe watcher, use sync.Cond instead of selectCase

### LINKED ISSUE #7495: Querynode addQueryChannel before watchDMLChannels
#### Steps/Code to reproduce:
Querynode addQueryChannel before watchDMLChannels, and queryCollection would register empty tSafeWatcher, because there is no vChannels in collection before calling watchDMLChannels.

#### Expected result:
search passed

#### Actual results:
search timeout
https://ci.milvus.io:18080/jenkins/blue/organizations/jenkins/milvus-ha-ci/detail/PR-7434/9/pipeline

#### Environment:
- Milvus version(e.g. v2.0.0-RC2 or 8b23a93):
- Deployment mode(standalone or cluster):
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:

### PR DESCRIPTION
**Pull request by bigsheeper**
Signed-off-by: bigsheeper <yihao.dai@zilliz.com>

Resolves: #7495

### CODE REVIEW

File: internal/querynode/query_collection.go [hunk 1]
```diff
@@ -140,19 +139,32 @@ func (q *queryCollection) register() {
...
+		zap.Any("collectionID", q.collectionID),
+		zap.Any("channel", vChannel),
+	)
+	go q.startWatcher(vChannel)
+}
+
+// TODO: add stopWatcher(), add close() to tSafeWatcher
+func (q *queryCollection) startWatcher(vChannel Channel) {
+	for {
+		// TODO: check if channel is closed
```
Comment (xige-16) [inline internal/querynode/query_collection.go:165]: This place will cause go routine leaks
suggestion:
```
	for {
                select {
                case <-ctx.Done():
			return
		case <-q.tSafeWatchers[vChannel].watcherChan():
                         q.watcherSharedChan <- struct{}{}
		// TODO: check if channel is closed
                }
	}
```

Comment (bigsheeper) [reply internal/querynode/query_collection.go]: fixed

File: internal/querynode/query_collection.go [hunk 2]
```diff
@@ -140,19 +140,45 @@ func (q *queryCollection) register() {
 		return
 	}
 
-	//TODO:: can't add new vChannel to selectCase
-	q.watcherSelectCase = make([]reflect.SelectCase, 0)
+	q.watcherSharedChan = make(chan struct{}, 1024)
```
Comment (xige-16) [inline internal/querynode/query_collection.go:143]: will block query if add query collection before watchDmChannel

Comment (bigsheeper) [reply internal/querynode/query_collection.go]: fixed

File: internal/querynode/tsafe_replica.go [hunk 3]
```diff
@@ -23,38 +23,44 @@ import (
...
+	removeTSafe(vChannel Channel) error
+	registerTSafeWatcher(vChannel Channel, watcher *tSafeWatcher) error
+	removeRecord(vChannel Channel, id UniqueID) error
 }
 
 type tSafeReplica struct {
-	mu     sync.Mutex        // guards tSafes
-	tSafes map[string]tSafer // map[vChannel]tSafer
+	mu         sync.Mutex        // guards tSafes
+	tSafes     map[string]tSafer // map[vChannel]tSafer
```
Comment (xiaofan-luan) [inline internal/querynode/tsafe_replica.go:36]: why tSafe use string as key but refCounter use vChannel as key

File: internal/querynode/task.go [hunk 4]
```diff
@@ -236,6 +236,13 @@ func (w *watchDmChannelsTask) Execute(ctx context.Context) error {
 		log.Debug("query node add collection flow graphs", zap.Any("channels", vChannels))
 	}
 
+	// add tSafe watcher if queryCollection exists
+	if w.node.queryService.hasQueryCollection(collectionID) {
+		for _, channel := range vChannels {
+			w.node.queryService.queryCollections[collectionID].addTSafeWatcher(channel)
```
Comment (xiaofan-luan) [inline internal/querynode/task.go:242]: what happens if queryCollections already exist,  is there any exception thrown out?

File: internal/querynode/tsafe_replica.go (hunk 3 above)
Comment (xiaofan-luan) [reply internal/querynode/tsafe_replica.go:36]: if we make tSafes and refCounter in one map that would be better, like

type TSaferRef struct {
    a tSafer,
    ref int
}
map[Channel] Pair

**Review by xiaofan-luan — COMMENTED** [review]
approve

**Pull request merged**